### PR TITLE
Refactoring MSD and implement NLPP parameter derivatives in evaluateDerivRatios

### PR DESCRIPTION
**Pull request by ye-luo**
## Proposed changes
Minimize code duplication and implement evaluateDerivRatios

## What type(s) of changes does this code introduce?
- New feature
- Testing changes (e.g. new unit/integration/performance tests)

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
epyc-server

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed

### CODE REVIEW

**Comment by ye-luo** [comment]
Test this please

**Comment by ye-luo** [comment]
Test this please

File: src/QMCWaveFunctions/tests/test_multi_slater_determinant.cpp [hunk 1]
```diff
@@ -179,6 +179,16 @@ void test_LiH_msd(const std::string& spo_xml_string,
 
     CHECK(std::real(ratios2[0]) == Approx(-0.8544310407));
     CHECK(std::real(ratios2[1]) == Approx(-1.0830708458));
+
+    std::fill(ratios2.begin(), ratios2.end(), 0);
+    Matrix<ValueType> dratio(2, nparam);
+    twf.evaluateDerivRatios(VP, active, ratios2, dratio);
+
+    CHECK(std::real(ratios2[0]) == Approx(-0.8544310407));
```
Comment (prckent) [inline src/QMCWaveFunctions/tests/test_multi_slater_determinant.cpp:187]: Where did these values come from?

File: tests/molecules/C2_pp/CMakeLists.txt [hunk 2]
```diff
@@ -77,6 +77,31 @@ if(NOT QMC_CUDA)
       C2_PP_MSDJ_LONG # VMC
     )
 
+    if(NOT QMC_MIXED_PRECISION)
+      list(APPEND DET_C2_PP_MSDJ_PARAM CIcoeff_4 0.29045551 0.00001)
+      list(APPEND DET_C2_PP_MSDJ_PARAM ud_4 0.4797811594 0.00001)
```
Comment (prckent) [inline tests/molecules/C2_pp/CMakeLists.txt:82]: Where did these values come from? Are they the current reported values or the result of some independent evaluation?

**Review by prckent — CHANGES_REQUESTED** [review]
Just minor requests re: How do we know that the new implementation is good?

File: src/QMCWaveFunctions/tests/test_multi_slater_determinant.cpp (hunk 1 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/tests/test_multi_slater_determinant.cpp:187]: ratios2 should be identical to the return of evaluateRatios call above.
dratio value is obtained from the old code path by accepting a move and then calling evaluateDeriv.

**Review by prckent — APPROVED** [review]

**Pull request merged**